app/services/projects/concerns/set_calculated_custom_field_values.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Projects::Concerns
+  module SetCalculatedCustomFieldValues
+    private
+
+    def set_attributes(params)
+      super(params_except_calculated_fields(params)).tap do
+        update_calculated_value_custom_fields
+      end
+    end
+
+    def params_except_calculated_fields(params)
+      custom_field_value_params = params[:custom_field_values]
+      return params unless custom_field_value_params
+
+      calculated_field_ids = model.all_available_custom_fields.select(&:field_format_calculated_value?).map(&:id)
```

Comment:
This call iterates the returned collection twice. We could instead leverage the database for filtering the format and only select the column we are interested in:

```suggestion
      calculated_field_ids = model.all_available_custom_fields.where(field_format: "calculated_value").pluck(:id)
```